pennylane/device_tracker.py

```diff
@@ -0,0 +1,171 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+from collections import defaultdict
+
+import pennylane as qml
+
+
+def track(dev, version="default", **kwargs):
+    r"""Creates a tracking context and applies it to a device.
+
+    Args:
+        dev (~.Device): a PennyLane-compatible device
+        version (str): name of tracker to use.  The current options are
+            `default` and `timing`.
+
+    Keyword Args:
+        reset_on_enter=True (bool): whether or not to reset information
+            entering the context
+
+    **Usage Information**
+
+    Note that with backpropagation, this functions should take ``qnode.device``
+    instead of the device used to create the QNode.
+
+    .. code-block:: python
+
+        dev = qml.device('default.qubit', wires=1)
+
+        @qml.qnode(dev, diff_method="parameter-shift")
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+    With the default version, total execution information is printed on
+    each device execution.  The printed data depends on the device and tracker version,
+    but for standard PennyLane devices, the object will track executions and shots.
+
+    >>> with qml.track(circuit.device) as tracker:
+    ...    qml.grad(circuit)(0.1, shots=10)
+    Total: executions = 1	shots = 10
+    Total: executions = 2	shots = 20
+    Total: executions = 3	shots = 30
+
+    In with the ``'timing'`` implementation, the instance also tracks the time
+    between entering the context and the completion of an execution.
+
+    >>> with qml.track(circuit.device, version='timing') as timing_tracker:
+    ...    circuit(0.1)
+    ...    circuit(0.2)
+    Total: executions = 1	time = 0.0011134147644042969
+    Total: executions = 2	time = 0.0027322769165039062
+
+    After completion, one can also access the recorded information:
+
+    >>> timing_tracker.totals
+    defaultdict(int, {'executions': 2, 'shots': 30, 'time': 0.00311279296875})
+    >>> timing_tracker.history
+    defaultdict(list,
+            {'executions': [1, 1],
+             'shots': [None, None],
+             'time': [0.0012764930725097656, 0.0018362998962402344]})
```

Comment:
This part is cool!

Reply:
Though do we need to use `defaultdict`?

Reply:
I didn't want to write
```
self.history[key] = self.history.get(key, []).append(value)
self.totals[key] = value + self.totals.get(key, 0)
```

But I guess that's not too bad.